File: first_time/DFS_BFS/130.被围绕的区域.py
#
# @lc app=leetcode.cn id=130 lang=python3
#
# [130] 被围绕的区域
#

# @lc code=start
from typing import List
class Solution:
    def __init__(self) -> None:
        self.dys = [-1, 1, 0, 0]
        self.dxs = [0, 0, -1, 1]
    
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        num_rows, num_cols = len(board), len(board[0])
        connected_o = [[0 for _ in range(num_cols)] for _ in range(num_rows)]
        board_indexs = []
        for x in range(num_cols):
            board_indexs.append((0, x))
            board_indexs.append((num_rows-1, x))
        for y in range(1, num_rows-1):
            board_indexs.append((y, 0))
            board_indexs.append((y, num_cols-1))
        for index in board_indexs:
            if board[index[0]][index[1]] == 'O':
                self._dfs(index[0], index[1], board, connected_o, num_rows, num_cols)
        
        for y in range(num_rows):
            for x in range(num_cols):
                if board[y][x] == 'O' and connected_o[y][x] == 0:
                    board[y][x] = 'X'

        


    def _dfs(self, y, x, board, connected_o, num_rows, num_cols):
        connected_o[y][x] = 1
        for dy, dx in zip(self.dys, self.dxs):
            if 0 <= y + dy < num_rows and 0 <= x+dx < num_cols:
                if board[y+dy][x+dx] == 'O' and connected_o[y+dy][x+dx] != 1:
                    connected_o[y+dy][x+dx] = 1
                    self._dfs(y+dy, x+dx, board, connected_o, num_rows, num_cols)
# @lc code=end


Solution().solve(
    [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
)

File: 130.被围绕的区域.cpp
/*
 * @lc app=leetcode.cn id=130 lang=cpp
 *
 * [130] 被围绕的区域
 */

#include <iostream>
#include <vector>
using namespace std;

// @lc code=start

class Solution {
public:
    void solve(vector<vector<char>>& board) {
        int m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for (int i=0;i<m;i++){
            if (board[i][0] == 'O'){
                visited[i][0] = true;
                dfs(board, visited, 0, i, n, m);
            }
            if (board[i][n-1] == 'O'){
                visited[i][n-1] = true;
                dfs(board, visited, n-1, i, n, m);
            }
        }

        for (int i=1;i<n-1;i++){
            if (board[0][i] == 'O'){
                visited[0][i] = true;
                dfs(board, visited, i, 0, n, m);
            }
            if (board[m-1][i] == 'O'){
                visited[m-1][i] = true;
                dfs(board, visited, i, m-1, n, m);
            }
        }
    
        for (int y=0;y<m;y++){
            for (int x=0;x<n;x++){
                if(!visited[y][x] && board[y][x]=='O'){
                    board[y][x] = 'X';
                }
            }
        }

    }

    void dfs(vector<vector<char>>& board, vector<vector<bool>>& visited, int x, int y, int n, int m){
        vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for (int i = 0; i < 4; i++){
            int next_x = x + directions[i].first;
            int next_y = y + directions[i].second;
            if (next_x < 0 || next_x >= n || next_y < 0 || next_y >= m) continue;
            if  (board[next_y][next_x] == 'O' && !visited[next_y][next_x]){
                visited[next_y][next_x] = true;
                dfs(board, visited, next_x, next_y, n, m);
            }
        }

    }
};
// @lc code=end

